segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java

```diff
@@ -105,6 +106,48 @@ public void close() {
 
     //region Operations
 
+    /**
+     * Executes the given action only if the Segment is empty.
+     *
+     * This is a conditional operation that will execute serially after all currently executing conditional updates on
+     * the given Segment have completed. Any subsequent conditional updates on the given Segment will be blocked until
+     * this action completes executing. Unconditional updates will not block the execution of this action, nor will they
+     * be blocked by its execution.
+     *
+     * Similarly to conditional updates, this can be used to ensure consistency across the TableSegment, making sure that
+     * the action only executes if a specific condition is met, while making sure no other consistency-related operations
+     * may begin while it is running.
+     *
+     * A deadlock will occur if action is used to perform a conditional update on the same segment. If an update is
+     * required, then an unconditional update should be executed via action.
+     *
+     * @param segment The Segment to perform the action on.
+     * @param action  A Supplier that, when invoked, will begin executing the action. This returns a CompletableFuture
+     *                that will indicate when the action completed.
+     * @param <T>     Return type.
+     * @return A CompletableFuture that, when completed normally, will indicate that the action was executed. If it failed,
+     * or if the segment is not empty, it will be failed with the appropriate exception. Notable exceptions:
+     * <ul>
+     * <li>{@link TableSegmentNotEmptyException} If the Segment is not empty.
+     * </ul>
+     */
+    <T> CompletableFuture<T> executeIfEmpty(DirectSegmentAccess segment, Supplier<CompletableFuture<T>> action) {
+        return this.recoveryTracker.waitIfNeeded(segment, () -> this.conditionalUpdateProcessor.addRange(
+                conditionKey -> conditionKey.getKey() == segment.getSegmentId(),
```

Comment:
Is this condition supposed to match all keys for the segment? I'm asking because I'm not sure I understand why `conditionKey.getKey() == segment.getSegmentId()` is supposed to evaluate to true for all keys. Could you clarify, please?

Reply:
I believe I explained this in the Javadoc for this method. `delete(empty==true)` is a conditional delete, conditioned on the segment being empty. As with any other conditional update, it needs to be sequenced behind all other conditional updates that affect the keys involved. Since this will affect all keys, it will need to wait for all conditional updates on all keys in the segment (and block subsequent conditional updates on that segment until it is done).

Reply:
I get that, but that's not what I'm asking. Let me elaborate.

The first parameter of `addRange` is a predicate for keys. If a key belongs, then it evaluates to true and otherwise false. Now, `conditionKey` is the argument representing a key we want to test and it will evaluate to true in the case that `conditionKey.getKey() == segment.getSegmentId()`. I didn't understand why `conditionKey.getKey() ` is supposed to evaluate to segment id, but I just sorted out my own question by looking at the definition in `ContainerKeyIndex`. `getKey()` confused me in this context.